functions/slack/SlackKnowledgeGraphSearch/Function.cs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Kgsearch.v1;
+using Google.Apis.Kgsearch.v1.Data;
+using Google.Apis.Services;
+using Google.Cloud.Functions.Framework;
+using Google.Cloud.Functions.Invoker;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+[assembly: FunctionsStartup(typeof(SlackKnowledgeGraphSearch.Startup))]
+
+namespace SlackKnowledgeGraphSearch
+{
+    // [START functions_slack_setup]
+    public class Startup : FunctionsStartup
+    {
+        private static readonly TimeSpan SlackTimestampTolerance = TimeSpan.FromMinutes(5);
+
+        public override void ConfigureServices(WebHostBuilderContext context, IServiceCollection services)
+        {
+            // These can come from an environment variable or a configuration file.
+            string kgApiKey = context.Configuration["KG_API_KEY"];
+            string slackSigningSecret = context.Configuration["SLACK_SECRET"];
+
+            services.AddSingleton(new KgsearchService(new BaseClientService.Initializer
+            {
+                ApiKey = kgApiKey
+            }));
+            services.AddSingleton(new SlackRequestVerifier(slackSigningSecret, SlackTimestampTolerance));
+            base.ConfigureServices(context, services);
+        }
+    }
+    // [END functions_slack_setup]
+
+    public class Function : IHttpFunction
+    {
+        private readonly ILogger _logger;
+        private readonly KgsearchService _kgService;
+        private readonly SlackRequestVerifier _verifier;
+
+        public Function(ILogger<Function> logger, KgsearchService kgService, SlackRequestVerifier verifier) =>
+            (_logger, _kgService, _verifier) = (logger, kgService, verifier);
```

Comment:
I would include these as well.

Reply:
I was wondering about that, but I wasn't sure where to include them. Ideally they'd be in the setup, but it's hard to show that.

For now I'll move things up so they're included in functions_slack_search. It won't include the class declaration, but it'll have the fields, the constructor, and the HandleAsync method.